crates/matrix-sdk-base/src/rooms/normal.rs

```diff
@@ -1230,12 +1230,56 @@ impl Room {
 
         self.update_seen_knock_request_ids(current_seen_events_guard, current_seen_events).await?;
 
-        self.store
-            .set_kv_data(
-                StateStoreDataKey::SeenKnockRequests(self.room_id()),
-                StateStoreDataValue::SeenKnockRequests(current_seen_events),
-            )
-            .await?;
+        Ok(())
+    }
+
+    /// Removes the seen knock request ids that are no longer valid given the
+    /// current room members.
+    pub async fn remove_outdated_seen_knock_requests_ids(&self) -> StoreResult<()> {
+        let current_seen_events_guard = self.get_write_guarded_current_knock_request_ids().await?;
+        // We're not calling `get_seen_join_request_ids` here because we need to keep
+        // the Mutex's guard until we've updated the data
+        let mut current_seen_events = current_seen_events_guard.clone().unwrap_or_default();
+
+        // Get and deserialize the member events for the seen knock requests
+        let keys: Vec<OwnedUserId> = current_seen_events.values().map(|id| id.to_owned()).collect();
+        let raw_member_events: Vec<RawMemberEvent> =
+            self.store.get_state_events_for_keys_static(self.room_id(), &keys).await?;
+        let member_events = raw_member_events
+            .into_iter()
+            .map(|raw| raw.deserialize())
+            .collect::<Result<Vec<MemberEvent>, _>>()?;
```

Comment:
It's a bit unfortunate that we have to load all the member events from the database systematically, only to filter them out later. The base crate has `BaseRoom::get_member(user_id)` and `StateStore::get_member_event(user_id)`, so we can do better. In pseudo-code:

```
for each (event_id, user_id) in current_seen_events {
  let member = store.get_member(room_id, user_id);
  if member.state is not knocked {
    remove it from current_seen_events
  }
}
```

And that way, we don't have to preload all the member events from the database. Does it make sense?

Reply:
But wouldn't that mean we now need to do N queries instead of a single one? I had that code before (with `store.get_member_event`) and I thought it wasn't as performant as it could be:
- With `store.get_member_event` you make several queries.
- With `room.get_member` you make several queries and also load the presence and other data that's not necessary at this point.

Reply:
> The keys passed are the user ids of the seen knock requests, so I don't think we're doing any extra work vs loading the members one by one.

Alright, I missed that, then we should be good to go :+1: